test/FunctionalTests/Microsoft.OData.Core.Tests/AsyncYieldStream.cs

```diff
@@ -0,0 +1,125 @@
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.OData.Tests
+{
+    /// <summary>
+    /// Stream wrapper that yields in every async method
+    /// to ensure the read/write operations complete asynchronously
+    /// even when the operate completely in-memory. This helps simulate
+    /// async I/O in unit tests without making requests out of the process.
+    /// </summary>
+    internal class AsyncYieldStream : Stream
+    {
+        private Stream stream;
+        public AsyncYieldStream(Stream stream)
+        {
+            this.stream = stream;
+        }
+
+        public override bool CanRead => this.stream.CanRead;
+
+        public override bool CanSeek => this.stream.CanSeek;
+
+        public override bool CanWrite => this.stream.CanWrite;
+
+        public override long Length => this.stream.Length;
+
+        public override int ReadTimeout { get => this.stream.ReadTimeout; set => this.stream.ReadTimeout = value; }
+
+        public override int WriteTimeout { get => this.stream.WriteTimeout; set => this.stream.WriteTimeout = value; }
+
+        public override bool CanTimeout => this.stream.CanTimeout;
+
+        public override void Close()
+        {
+            this.stream.Close();
+        }
+
+        public override long Position { get => this.stream.Position; set => this.stream.Position = value; }
+
+        public override void Flush()
+        {
+            this.stream.FlushAsync().Wait();
+        }
+
+        public override async Task FlushAsync(CancellationToken cancellationToken)
+        {
+            await Task.Yield();
+            await base.FlushAsync(cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return this.stream.ReadAsync(buffer, offset, count).Result;
+        }
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return this.stream.ReadAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override int ReadByte()
+        {
+            return this.stream.ReadByte();
+        }
+
+        public override void WriteByte(byte value)
+        {
+            this.stream.WriteByte(value);
+        }
+
+        public override Task CopyToAsync(Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            return this.stream.CopyToAsync(destination, bufferSize, cancellationToken);
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            return this.stream.Seek(offset, origin);
+        }
+
+        public override void SetLength(long value)
+        {
+            this.stream.SetLength(value);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            this.stream.WriteAsync(buffer, offset, count).GetAwaiter().GetResult();

```

Comment:
I think you should call the method in `this.stream` that matches the signature you're overriding instead of mixing sync and async methods.